### ISSUE TITLE
Support for row_to_json, json_agg, and coalesce to support direct Postgres -> JSON encoding

### ISSUE THREAD

**Issue by njaremko** [tomjaguarpaw/haskell-opaleye#497]
Does Opaleye support `row_to_json` and `json_agg`? I haven't been able to find it in the docs, but figured I'd ask. If it doesn't, would there be interest in adding it? [Hasura](https://github.com/hasura/graphql-engine) is a haskell project that compiles GraphQL queries into Postgres queries that use the aforementioned functions to have the DB do the JSON encoding (as opposed to querying the DB, parsing the query results, and re-encoding it as JSON) 

An example query looks something like this:
```sql
SELECT
  coalesce(json_agg("root"), '[]') AS "root"
FROM
  (
    SELECT
      row_to_json(
        (
          SELECT
            "_4_e"
          FROM
            (
              SELECT
                "_3_root.or.support_ticket_type"."support_ticket_type" AS "support_ticket_type",
                "_0_root.base"."summary" AS "summary",
                "_0_root.base"."details" AS "details"
            ) AS "_4_e"
        )
      ) AS "root"
    FROM
      (
        SELECT
          *
        FROM
          "public"."support_tickets"
        WHERE
          ('true')
      ) AS "_0_root.base"
      LEFT OUTER JOIN LATERAL (
        SELECT
          row_to_json(
            (
              SELECT
                "_2_e"
              FROM
                (
                  SELECT
                    "_1_root.or.support_ticket_type.base"."id" AS "id",
                    "_1_root.or.support_ticket_type.base"."name" AS "name"
                ) AS "_2_e"
            )
          ) AS "support_ticket_type"
        FROM
          (
            SELECT
              *
            FROM
              "public"."support_ticket_types"
            WHERE
              (
                ("_0_root.base"."support_ticket_type_id") = ("id")
              )
          ) AS "_1_root.or.support_ticket_type.base"
      ) AS "_3_root.or.support_ticket_type" ON ('true')
  ) AS "_5_root"
```

which returns JSON of the form:
```json
[{"support_ticket_type":{"id":1,"name":"....."},"summary":".....","details":"......"}, ...]
```

**Comment by tomjaguarpaw** [comment]
### `json_agg`

[Documentation](https://www.postgresql.org/docs/9.5/functions-aggregate.html)

Should be easy to add along the same lines as [`arrayAgg`](https://github.com/tomjaguarpaw/haskell-opaleye/blob/b7aacc07c6392654cae439fc3b997620c3aa7a87/src/Opaleye/Aggregate.hs#L143-L144). I'd welcome a PR for this. If you have any difficulty working out how to do this then please come back to me.

### `row_to_json`

[Documentation](https://www.postgresql.org/docs/9.5/functions-json.html)

Trickier to add. Will probably need a `ProductProfunctor` defined for a simple API.
If you want to try to tackle this then you are welcome but I don't think it is for the faint of heart!

**Comment by njaremko** [comment]
If they're welcome, then I'll get started on some PRs 👍

**Comment by tomjaguarpaw** [comment]
Definitely welcome!

**Comment by njaremko** [comment]
I've opened #498 to add `json_agg`, though I need a bit of guidance from you with regards to testing (I explain in more detail in the PR).

Regarding `row_to_json`, you mentioned that a nice API would probably require a `ProductProfunctor` definition. Can you give an example of the API you had in mind? Then I can take a stab at implementing it?

**Comment by tomjaguarpaw** [comment]
I think I'm perhaps misunderstanding how these operations are used in practice. Could you give me a small example of SQL, and also imaginary Opaleye code that demonstrates how you would like to use them?

**Comment by njaremko** [comment]
I haven't given enough thought to how I want the syntax to look yet, but I can provide some examples of using these functions. Using tables from the test suite, here are some sample queries:

`json_agg` takes either a row, or a subquery as it's argument. If you give a row, you'll get useless column names, so generally a subquery is used.

# Row example

```sql
SELECT json_agg((column1, column2)) FROM table6
```
returns
```json
[{"f1":"xy","f2":"a"}, 
 {"f1":"z","f2":"a"}, 
 {"f1":"more text","f2":"a"}]
```

# Subquery example

```sql
SELECT json_agg(t) FROM (SELECT * FROM table6) as t
```
returns
```json
[{"column1":"xy","column2":"a"}, 
 {"column1":"z","column2":"a"}, 
 {"column1":"more text","column2":"a"}]
```

# `row_to_json` example
### Note that it behaves similarly to json_agg above, accepting a row, or a sub-query. Using a row loses column information here as well.
```sql
SELECT row_to_json(t) FROM (SELECT column1, column2 from table6) as t
```
returns (a column with three rows)
```
{"column1":"xy","column2":"a"}
{"column1":"z","column2":"a"}
{"column1":"more text","column2":"a"}
```

# Use with `json_build_object`
This function allows you to specify the field names in the function call, which might be easier to implement in Opaleye (I'll defer to you on this). However, it's a bit slower than `row_to_json`.

```sql
SELECT json_agg(json_build_object('column1', column1, 'column2', column2)) FROM table6
```
returns
```json
[{"column1" : "xy", "column2" : "a"}, {"column1" : "z", "column2" : "a"}, {"column1" : "more text", "column2" : "a"}]
```

# More realistic use
### Note that we're aliasing the columns here to change the json fields
```sql
SELECT coalesce(json_agg("root"), '[]')
FROM (
    SELECT row_to_json("t1") AS "root"
    FROM (SELECT column1 as apples, column2 as fish from table6) as "t1"
    ) as "t2";
```
returns 
```json
[{"apples":"xy","fish":"a"}, {"apples":"z","fish":"a"}, {"apples":"more text","fish":"a"}]
```

# Longer example (with joins)
### Note the use of aliasing to rename columns in the returned json
```sql
SELECT coalesce(json_agg("root"), '[]')
FROM (
    SELECT row_to_json(
        (
            SELECT t6
            FROM (
                SELECT
                       "table1.base".column1 as "id",
                       "root.or.TABLE2"."TABLE2_stuff" as "blog_post"
                ) as t6
        )
    ) AS "root"
    FROM (
        SELECT * from table1
        ) as "table1.base"
    LEFT OUTER JOIN LATERAL (
        SELECT
          row_to_json(
            (
                SELECT "t4"
                FROM
                  (
                    SELECT
                        "TABLE2.base".column1 as "summary",
                        "TABLE2.base".column2 as "details"
                  ) as "t4"
            )
        ) as "TABLE2_stuff"
        FROM
          (
            SELECT * FROM "TABLE2"
              where "table1.base".column1 = column1
          ) as "TABLE2.base"
        ) as "root.or.TABLE2" ON ('true')
    ) as "t5_root";
```
returns
```json
[{"id":1,"blog_post":{"summary":1,"details":100}}, {"id":1,"blog_post":{"summary":1,"details":100}}, {"id":1,"blog_post":{"summary":1,"details":100}}, {"id":2,"blog_post":null}]
```

**Comment by tomjaguarpaw** [comment]
Aha, thanks for the detailed examples.

Using column names isn't going to work well because Opaleye makes up arbitrary unique identifiers for column names, things like `binary_357`. On the other hand it will be possible to support `json_build_object` because the column names are provided explicitly.

Would `json_build_object` plus `json_agg` of a single column suit your use case? If so then I can show you how to do the former and your PR #498 will be suitable for the latter.

If subsequently you find `json_build_object` too slow then we can investigate methods of speeding it up, but I think it would be good to get something working in the first place.

**Comment by njaremko** [comment]
I'm certainly fine with getting this workflow working using `json_build_object`, the only issue with that approach is that we're capped at 50 fields (100 arguments), but that should be fine in most cases...

When you say "`json_agg` of a single column" are you referring to my example above? `json_agg(json_build_object(.....))`?

Could I build a query like the following?
```sql
SELECT coalesce(
    json_agg(
        json_build_object(
            'id', "table1.base".column1, 
            'blog_post', "root.or.TABLE2"."TABLE2_stuff"
        )
    ), '[]')
FROM 
    (
        SELECT *
        FROM table1
    ) AS "table1.base"
LEFT OUTER JOIN LATERAL 
    (
        SELECT json_build_object(
            'summary', "TABLE2.base".column1, 
            'details', "TABLE2.base".column2
        ) AS "TABLE2_stuff"
        FROM 
        (
             SELECT *
             FROM "TABLE2"
             WHERE "table1.base".column1 = column1
        ) AS "TABLE2.base"
    ) AS "root.or.TABLE2" ON ('true')
```

 If so, then yes, that sounds like it would work fine.

**Comment by tomjaguarpaw** [comment]
Putting sub-queries inside `json_build_object` will be tricky, but columns should be fine. Here are some hints about how to do that:

Make a datatype that represents JSON field/value pairs and hide the constructor (i.e. only make it available in a `Opaleye.Internal` module).

```haskell
data JSONBuildObjectFields =
    JSONBuildObjectFields [(String, PrimExpr)]

instance Monoid JSONBuildObjectFields where
    ...

jsonBuildObjectField :: String -> Column a -> JSONBuildObjectFields
jsonBuildObjectField f v = [(f, v)]

jsonBuildObject :: JSONBuildObjectFields -> Column SqlJson
jsonBuildObject jbofs = FunExpr "json_build_object" args
    where args = concatMap ... jbofs
```

If you have trouble working out how to fill in `...` then let me know and I'll explain in more detail when I have a bit more time.

**Comment by njaremko** [comment]
I've pushed an implementation of where I think you were going with this to #498 . We can probably move conversation there now?

**Comment by njaremko** [comment]
This has been implemented in #498 .
